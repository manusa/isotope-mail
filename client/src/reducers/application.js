import {INITIAL_STATE} from './';
import {ActionTypes} from '../actions/action-types';

const application = (state = INITIAL_STATE.application, action = {}) => {
  switch (action.type) {
    case ActionTypes.APPLICATION_BE_REQUEST:
      return {...state, activeRequests: state.activeRequests + 1};
    case ActionTypes.APPLICATION_BE_REQUEST_COMPLETED:
      return {...state, activeRequests: state.activeRequests > 0 ? state.activeRequests - 1 : 0};
    case ActionTypes.APPLICATION_USER_CREDENTIALS_SET:
      return {
        ...state, user: {
          ...state.user,
          id: action.payload.userId, hash: action.payload.hash, credentials: action.payload.credentials
        }
      };
    case ActionTypes.APPLICATION_FOLDER_SELECT:
      return {
        ...state,
        selectedFolderId: action.payload.folderId
      };
    case ActionTypes.APPLICATION_FOLDER_RENAME: {
      return {
        ...state,
        renameFolderId: action.payload ? action.payload.folderId : null
      };
    }
    case ActionTypes.APPLICATION_FOLDER_RENAME_OK: {
      const newState = {...state};
      if (state.selectedFolderId === action.payload.oldFolderId) {
        newState.selectedFolderId = action.payload.newFolderId;
      }
      return newState;
    }
    case ActionTypes.APPLICATION_MESSAGE_SELECT:
      return {...state, selectedMessage: {...action.payload}};
    case ActionTypes.APPLICATION_MESSAGE_REFRESH: {
      const newState = {...state};
      const message = action.payload.message;
      // Store in application.downloadedMessages
      newState.downloadedMessages[message.messageId] = message;
      // Update selected message if is currently selected
      if (state.selectedFolderId === action.payload.folder.folderId
        && state.selectedMessage.uid === message.uid) {
        newState.selectedMessage = {...message};
      }
      return newState;
    }
    case ActionTypes.APPLICATION_MESSAGE_REFRESH_BE_REQUEST:
      return {...state, refreshMessageActiveRequests: state.refreshMessageActiveRequests + 1};
    case ActionTypes.APPLICATION_MESSAGE_REFRESH_BE_REQUEST_COMPLETED:
      return {...state, refreshMessageActiveRequests:
          state.refreshMessageActiveRequests > 0 ? state.refreshMessageActiveRequests - 1 : 0};
    case ActionTypes.APPLICATION_MESSAGE_PRE_DOWNLOAD: {
      const newState = {...state};
      const messages = action.payload.messages;
      // Store in application.downloadedMessages
      messages.forEach(message => {
        newState.downloadedMessages[message.messageId] = message;
      });
      return newState;
    }
    case ActionTypes.APPLICATION_MESSAGE_REPLACE_IMAGE: {
      const newState = {...state};
      const folder = action.payload.folder;
      const message = action.payload.message;
      const attachment = action.payload.attachment;
      const contentId = attachment.contentId.replace(/[<>]/g, '');
      const regex = new RegExp(`cid:${contentId}`, 'g'); // Multiple occurrence -> ReplaceAll
      if (contentId.length > 0) {
        const objectUrl = URL.createObjectURL(action.payload.blob);
        // Store in application.downloadedMessages
        newState.downloadedMessages[message.messageId].content =
          newState.downloadedMessages[message.messageId].content.replace(regex, objectUrl);
        // Update selected message if applicable
        if (newState.selectedFolderId === folder.folderId
          && newState.selectedMessage.uid === message.uid) {
          const parsedMessage = newState.selectedMessage.content.replace(regex, objectUrl);
          newState.selectedMessage = {...newState.selectedMessage, content: parsedMessage};
        }
      }
      return newState;
    }
    case ActionTypes.APPLICATION_ERROR_SET:
      const errorsSetState = {...state};
      errorsSetState.errors = {...state.errors};
      errorsSetState.errors[action.payload.type] = action.payload.value;
      return errorsSetState;
    case ActionTypes.APPLICATION_MESSAGE_EDIT: {
      const newState = {...state};
      newState.newMessage = action.payload ? {...action.payload} : null;
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_SEND: {
      const newState = {...state};
      newState.outbox = {message: action.payload, progress: 0, sent: false};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_UPDATE_PROGRESS: {
      const newState = {...state};
      newState.outbox = {...newState.outbox, progress: action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_SET_SENT: {
      const newState = {...state};
      newState.outbox = {...newState.outbox, sent: action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_SET_ERROR: {
      const newState = {...state};
      newState.outbox = {...newState.outbox, error: action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_MESSAGE_PROCESSED: {
      const newState = {...state};
      newState.outbox = null;
      return newState;
    }
    default:
      return state;
  }
};

export default application;
