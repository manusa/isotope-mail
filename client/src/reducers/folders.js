import {initialState} from './';
import {ActionTypes} from '../actions/action-types';

const folders = (state = initialState.folders, action) => {
  switch (action.type) {
    case ActionTypes.ADD_FOLDER:
      // return {...state, folders: [...state.folders, action.payload]}
      return [...state, action.payload];
    default:
      return state;
  }
}

export default folders;
