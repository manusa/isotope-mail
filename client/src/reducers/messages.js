import {initialState} from './';
import {ActionTypes} from '../actions/action-types';

const messages = (state = initialState.messages, action) => {
  switch (action.type) {
    case ActionTypes.MESSAGES_BE_REQUEST:
      return {...state, activeRequests: state.activeRequests + 1};
    case ActionTypes.MESSAGES_BE_REQUEST_COMPLETED:
      return {...state, activeRequests: state.activeRequests > 0 ? state.activeRequests - 1 : 0};
    case ActionTypes.MESSAGES_UPDATE_CACHE:
      const newState = {...state};
      newState.cache[action.payload.folder.folderId] =
        new Map(action.payload.messages.map(m => [m.uid, m]));
      return newState;
    case ActionTypes.ADD_MESSAGE:
      return {...state, items: [...state.items, action.payload]};
    default:
      return state;
  }
};

export default messages;
