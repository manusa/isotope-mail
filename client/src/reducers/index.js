import {combineReducers} from 'redux';
import application from './application';
import folders from './folders';
import login from './login';
import messages from './messages';

export const INITIAL_STATE = {
  application: {
    title: 'Isotope Mail Client',
    user: {},
    newMessage: null,
    selectedFolderId: {},
    messageFilterKey: null,
    messageFilterText: '',
    createFolderParentId: null,
    renameFolderId: null,
    selectedMessage: null,
    downloadedMessages: {},
    outbox: null,
    pollInterval: 15000,
    errors: {
      diskQuotaExceeded: false,
      authentication: null
    },
    refreshMessageActiveRequests: 0,
    activeRequests: 0
  },
  folders: {
    items: [],
    explodedItems: {},
    activeRequests: 0
  },
  login: {
    formValues: {}
  },
  messages: {
    cache: {},
    selected: [],
    locked: [],
    activeRequests: 0
  }
};

// https://github.com/reduxjs/redux/issues/749#issuecomment-164327121
export default combineReducers({
  application,
  login,
  folders: folders,
  messages: messages
});
