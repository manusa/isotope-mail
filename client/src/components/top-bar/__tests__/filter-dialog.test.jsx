import React from 'react';
import {shallow} from 'enzyme';
import {createMockStore} from '../../../__testutils__/store';
import {INITIAL_STATE} from '../../../reducers';
import MessageFilters from '../../../services/message-filters';
import FilterDialogConnected, {FilterDialog} from '../filter-dialog';

describe('FilterDialog component test suite', () => {
  let commonProps;

  beforeEach(() => {
    commonProps = {
      t: jest.fn(arg => arg),
      closeFilterDialogHandler: jest.fn(),
      activeMessageFilter: MessageFilters.ALL,
      messageFilterText: '',
      setMessageFilterKey: jest.fn(),
      setMessageFilterText: jest.fn()
    };
  });

  describe('Snapshot render', () => {
    test('Snapshot render, visible, should render filter-dialog', () => {
      // Given
      const props = {...commonProps, visible: true};
      // When
      const filterDialog = shallow(<FilterDialog {...props}/>);
      // Then
      expect(filterDialog).toMatchSnapshot();
    });
    test('Snapshot render, visible=false, should render filter-dialog "hidden"', () => {
      // Given
      const props = {...commonProps, visible: false};
      // When
      const filterDialog = shallow(<FilterDialog {...props}/>);
      // Then
      expect(filterDialog).toMatchSnapshot();
    });
  });
  describe('Events tests', () => {
    test('Input text filter changed, should invoke setMessageFilterText', () => {
      // Given
      const props = {...commonProps, visible: true};
      const filterDialog = shallow(<FilterDialog {...props}/>);
      // When
      filterDialog.find('li').find('input').simulate('change', {target: {}});
      // Then
      expect(props.setMessageFilterText).toHaveBeenCalledTimes(1);
    });
    test('Input text filter mousedown, should disable closeFilterDialogHandler', () => {
      // Given
      const props = {...commonProps, visible: true};
      const filterDialog = shallow(<FilterDialog {...props}/>);
      // When
      filterDialog.find('li').find('input').simulate('mousedown', {target: {}});
      // Then
      expect(props.closeFilterDialogHandler.disabled).toBe(true);
    });
    test('Item click, should invoke setMessageFilterKey', () => {
      // Given
      const props = {...commonProps, visible: true};
      const filterDialog = shallow(<FilterDialog {...props}/>);
      // When
      filterDialog.find('li').findWhere(li => li.key() === 'ALL').simulate('click', {});
      // Then
      expect(props.setMessageFilterKey).toHaveBeenCalledTimes(1);
    });
  });
  describe('Redux properties', () => {
    test('setMessageFilterKey, should invoke application setMessageFilterKey action', () => {
      // Given
      const store = createMockStore({...INITIAL_STATE});
      const filterDialog = shallow(<FilterDialogConnected store={store} />);
      // When
      filterDialog.props().setMessageFilterKey(MessageFilters.READ);
      // Then
      expect(store.getState().application.messageFilterKey).toBe('READ');
    });
    test('setMessageFilterText, should invoke application setMessageFilterText action', () => {
      // Given
      const store = createMockStore({...INITIAL_STATE});
      const filterDialog = shallow(<FilterDialogConnected store={store} />);
      // When
      filterDialog.props().setMessageFilterText('1337');
      // Then
      expect(store.getState().application.messageFilterText).toBe('1337');
    });
  });
});
