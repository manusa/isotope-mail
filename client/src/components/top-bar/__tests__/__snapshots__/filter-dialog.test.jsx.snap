// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDialog component test suite Snapshot render Snapshot render, visible, should render filter-dialog 1`] = `
<div
  aria-hidden={false}
  className="filter-dialog mdc-menu mdc-menu-surface
    mdc-menu-surface--open"
>
  <ul
    className="mdc-list mdc-list--dense"
  >
    <li
      className="filter-dialog__item-text  mdc-list-item"
    >
      <input
        className="text-field"
        onChange={[Function]}
        onMouseDown={[Function]}
        value=""
      />
      <i
        className="icon material-icons mdc-button__icon"
      >
        search
      </i>
    </li>
    <li
      className="filter-dialog__item-check mdc-list-item
          mdc-list-item--selected"
      key="ALL"
      onClick={[Function]}
    >
      <i
        className="check check--active material-icons"
      >
        check
      </i>
      messageList.filters.All
    </li>
    <li
      className="filter-dialog__item-check mdc-list-item
          "
      key="READ"
      onClick={[Function]}
    >
      <i
        className="check  material-icons"
      >
        check
      </i>
      messageList.filters.Read
    </li>
    <li
      className="filter-dialog__item-check mdc-list-item
          "
      key="UNREAD"
      onClick={[Function]}
    >
      <i
        className="check  material-icons"
      >
        check
      </i>
      messageList.filters.Unread
    </li>
    <li
      className="filter-dialog__item-check mdc-list-item
          "
      key="FLAGGED"
      onClick={[Function]}
    >
      <i
        className="check  material-icons"
      >
        check
      </i>
      messageList.filters.Flagged
    </li>
  </ul>
</div>
`;

exports[`FilterDialog component test suite Snapshot render Snapshot render, visible=false, should render filter-dialog "hidden" 1`] = `
<div
  aria-hidden={true}
  className="filter-dialog mdc-menu mdc-menu-surface
    "
>
  <ul
    className="mdc-list mdc-list--dense"
  >
    <li
      className="filter-dialog__item-text  mdc-list-item"
    >
      <input
        className="text-field"
        onChange={[Function]}
        onMouseDown={[Function]}
        value=""
      />
      <i
        className="icon material-icons mdc-button__icon"
      >
        search
      </i>
    </li>
    <li
      className="filter-dialog__item-check mdc-list-item
          mdc-list-item--selected"
      key="ALL"
      onClick={[Function]}
    >
      <i
        className="check check--active material-icons"
      >
        check
      </i>
      messageList.filters.All
    </li>
    <li
      className="filter-dialog__item-check mdc-list-item
          "
      key="READ"
      onClick={[Function]}
    >
      <i
        className="check  material-icons"
      >
        check
      </i>
      messageList.filters.Read
    </li>
    <li
      className="filter-dialog__item-check mdc-list-item
          "
      key="UNREAD"
      onClick={[Function]}
    >
      <i
        className="check  material-icons"
      >
        check
      </i>
      messageList.filters.Unread
    </li>
    <li
      className="filter-dialog__item-check mdc-list-item
          "
      key="FLAGGED"
      onClick={[Function]}
    >
      <i
        className="check  material-icons"
      >
        check
      </i>
      messageList.filters.Flagged
    </li>
  </ul>
</div>
`;
